8341260: Add Float16 to jdk.incubator.vector

jddarcy opened the pull request:
Port of Float16 from java.lang in the lworld+fp16 branch to jdk.incubabor.vector.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8342567](https://bugs.openjdk.org/browse/JDK-8342567) to be approved

### Issues
 * [JDK-8341260](https://bugs.openjdk.org/browse/JDK-8341260): Add Float16 to jdk.incubator.vector (**Enhancement** - P4)
 * [JDK-8342567](https://bugs.openjdk.org/browse/JDK-8342567): Add Float16 to jdk.incubator.vector (**CSR**)


### Reviewers
 * [Raffaello Giulietti](https://openjdk.org/census#rgiulietti) (@rgiulietti - **Reviewer**)

### Contributors
 * Raffaello Giulietti `<rgiulietti@openjdk.org>`
 * Jatin Bhateja `<jbhateja@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21574/head:pull/21574` \
`$ git checkout pull/21574`

Update a local copy of the PR: \
`$ git checkout pull/21574` \
`$ git pull https://git.openjdk.org/jdk.git pull/21574/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21574`

View PR using the GUI difftool: \
`$ git pr show -t 21574`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21574.diff">https://git.openjdk.org/jdk/pull/21574.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21574#issuecomment-2424208732)
</details>

jddarcy commented:
/contributor add @rgiulietti

jddarcy commented:
/contributor add @jatin-bhateja

jddarcy commented:
Add as contributors other engineers who worked on Float16 on the lworld+fp16 branch in Valhalla.

jddarcy commented:
To ease review, diffs of corresponding files from the current, at time of writing, state of files in lworld+fp16 vs a port to jdk.incubator.vector, starting with Float16:

$ diff src/java.base/share/classes/java/lang/Float16.java  /home/darcy/JDK/Scm/git/jdk-full/open/src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Float16.java 
26c26
< package java.lang;
---
> package jdk.incubator.vector;
28a29
> import java.math.BigInteger;
30,31c31,32
< import jdk.internal.math.*;
< import jdk.internal.vm.annotation.IntrinsicCandidate;
---
> // import jdk.internal.math.*;
> // import jdk.internal.vm.annotation.IntrinsicCandidate;
37c38
<  * The {@code Float16} is a primitive value class holding 16-bit data
---
>  * The {@code Float16} is a class holding 16-bit data
46,48d46
<  * <p>This is a <a href="https://openjdk.org/jeps/401">primitive value class</a> and its objects are
<  * identity-less non-nullable value objects.
<  *
52a51,56
>  * <p>This is a <a href="{@docRoot}/java.base/java/lang/doc-files/ValueBased.html">value-based</a>
>  * class; programmers should treat instances that are
>  * {@linkplain #equals(Object) equal} as interchangeable and should not
>  * use instances for synchronization, or unpredictable behavior may
>  * occur. For example, in a future release, synchronization may fail.
>  *
62,64d65
<  *
<  * @author Jatin Bhateja
<  * @since 20.00
69,70c70,71
< @jdk.internal.MigratedValueClass
< @jdk.internal.ValueBased
---
> // @jdk.internal.MigratedValueClass
> //@jdk.internal.ValueBased
213c214,215
<         return FloatToDecimal.toString(f16.floatValue());
---
>         return Float.toString(f16.floatValue());
>         // return FloatToDecimal.toString(f16.floatValue());
420d421
<      * @see BigDecimal#float16Value()
423c424,539
<         return v.float16Value();
---
>         return BigDecimalConversion.float16Value(v);
>     }
> 
>     private class BigDecimalConversion {
>         /*
>          * Let l = log_2(10).
>          * Then, L < l < L + ulp(L) / 2, that is, L = roundTiesToEven(l).
>          */
>         private static final double L = 3.321928094887362;
> 
>         private static final int P_F16 = PRECISION;  // 11
>         private static final int Q_MIN_F16 = MIN_EXPONENT - (P_F16 - 1);  // -24
>         private static final int Q_MAX_F16 = MAX_EXPONENT - (P_F16 - 1);  // 5
> 
>         /**
>          * Powers of 10 which can be represented exactly in {@code
>          * Float16}.
>          */
>         private static final Float16[] FLOAT16_10_POW = {
>             Float16.valueOf(1), Float16.valueOf(10), Float16.valueOf(100),
>             Float16.valueOf(1_000), Float16.valueOf(10_000)
>         };
> 
>         public static Float16 float16Value(BigDecimal bd) {
> //             int scale = bd.scale();
> //             BigInteger unscaledValue = bd.unscaledValue();
> 
> //              if (unscaledValue.abs().compareTo(BigInteger.valueOf(Long.MAX_VALUE)) <= 0) {
> //                 long intCompact = bd.longValue();
> //                 Float16 v = Float16.valueOf(intCompact);
> //                 if (scale == 0) {
> //                     return v;
> //                 }
> //                 /*
> //                  * The discussion for the double case also applies here. That is,
> //                  * the following test is precise for all long values, but here
> //                  * Long.MAX_VALUE is not an issue.
> //                  */
> //                 if (v.longValue() == intCompact) {
> //                     if (0 < scale && scale < FLOAT16_10_POW.length) {
> //                         return Float16.divide(v, FLOAT16_10_POW[scale]);
> //                     }
> //                     if (0 > scale && scale > -FLOAT16_10_POW.length) {
> //                         return Float16.multiply(v, FLOAT16_10_POW[-scale]);
> //                     }
> //                 }
> //             }
>             return fullFloat16Value(bd);
>         }
> 
>         private static BigInteger bigTenToThe(int scale) {
>             return BigInteger.TEN.pow(scale);
>         }
> 
>         private static Float16 fullFloat16Value(BigDecimal bd) {
>             if (BigDecimal.ZERO.compareTo(bd) == 0) {
>                 return Float16.valueOf(0);
>             }
>             BigInteger w = bd.unscaledValue().abs();
>             int scale = bd.scale();
>             long qb = w.bitLength() - (long) Math.ceil(scale * L);
>             Float16 signum = Float16.valueOf(bd.signum());
>             if (qb < Q_MIN_F16 - 2) {  // qb < -26
>                 return Float16.multiply(signum, Float16.valueOf(0));
>             }
>             if (qb > Q_MAX_F16 + P_F16 + 1) {  // qb > 17
>                 return Float16.multiply(signum, Float16.POSITIVE_INFINITY);
>             }
>             if (scale < 0) {
>                 return Float16.multiply(signum, valueOf(w.multiply(bigTenToThe(-scale))));
>             }
>             if (scale == 0) {
>                 return Float16.multiply(signum, valueOf(w));
>             }
>             int ql = (int) qb - (P_F16 + 3);
>             BigInteger pow10 =  bigTenToThe(scale);
>             BigInteger m, n;
>             if (ql <= 0) {
>                 m = w.shiftLeft(-ql);
>                 n = pow10;
>             } else {
>                 m = w;
>                 n = pow10.shiftLeft(ql);
>             }
>             BigInteger[] qr = m.divideAndRemainder(n);
>             /*
>              * We have
>              *      2^12 = 2^{P+1} <= i < 2^{P+5} = 2^16
>              * Contrary to the double and float cases, where we use long and int, resp.,
>              * here we cannot simply declare i as short, because P + 5 < Short.SIZE
>              * fails to hold.
>              * Using int is safe, though.
>              *
>              * Further, as Math.scalb(Float16) does not exists, we fall back to
>              * Math.scalb(double).
>              */
>             int i = qr[0].intValue();
>             int sb = qr[1].signum();
>             int dq = (Integer.SIZE - (P_F16 + 2)) - Integer.numberOfLeadingZeros(i);
>             int eq = (Q_MIN_F16 - 2) - ql;
>             if (dq >= eq) {
>                 return Float16.valueOf(bd.signum() * Math.scalb((double) (i | sb), ql));
>             }
>             int mask = (1 << eq) - 1;
>             int j = i >> eq | (Integer.signum(i & mask)) | sb;
>             return Float16.valueOf(bd.signum() * Math.scalb((double) j, Q_MIN_F16 - 2));
>         }
> 
>         public static Float16 valueOf(BigInteger bi) {
>             int signum = bi.signum();
>             return (signum == 0 || bi.bitLength() <= 31)
>                 ? Float16.valueOf(bi.longValue())  // might return infinities
>                 : signum > 0
>                 ? Float16.POSITIVE_INFINITY
>                 : Float16.NEGATIVE_INFINITY;
>         }
577,578c693
<      * according to the IEEE 754 floating-point binary16 bit layout,
<      * preserving Not-a-Number (NaN) values.
---
>      * according to the IEEE 754 floating-point binary16 bit layout.
591,607d705
<      * Returns a representation of the specified floating-point value
<      * according to the IEEE 754 floating-point binary16 bit layout.
<      *
<      * @param   fp16   a {@code Float16} floating-point number.
<      * @return the bits that represent the floating-point number.
<      *
<      * @see Float#floatToIntBits(float)
<      * @see Double#doubleToLongBits(double)
<      */
<     public static short float16ToShortBits(Float16 fp16) {
<         if (!isNaN(fp16)) {
<             return float16ToRawShortBits(fp16);
<         }
<         return 0x7e00;
<     }
< 
<     /**
694c792
<     @IntrinsicCandidate
---
>     // @IntrinsicCandidate
714c812
<     @IntrinsicCandidate
---
>     // @IntrinsicCandidate
783c881
<     @IntrinsicCandidate
---
>     // @IntrinsicCandidate
806c904
<     @IntrinsicCandidate
---
>     // @IntrinsicCandidate
829c927
<     @IntrinsicCandidate
---
>     // @IntrinsicCandidate
852c950
<     @IntrinsicCandidate
---
>     // @IntrinsicCandidate
873c971
<     @IntrinsicCandidate
---
>     // @IntrinsicCandidate
907c1005
<     @IntrinsicCandidate
---
>     // @IntrinsicCandidate
1109c1207
<     @IntrinsicCandidate
---
>     // @IntrinsicCandidate
1131c1229
<     @IntrinsicCandidate
---
>     // @IntrinsicCandidate
1319a1418
>         int DoubleConsts_EXP_BIAS = 1023;
1328c1427
<                 * Double.longBitsToDouble((long) (scaleFactor + DoubleConsts.EXP_BIAS) << Double.PRECISION - 1));
---
>                 * Double.longBitsToDouble((long) (scaleFactor + DoubleConsts_EXP_BIAS) << Double.PRECISION - 1));
1330d1428
< 
1374d1471
<

jddarcy commented:
$ diff src/java.base/share/classes/jdk/internal/math/Float16Consts.java  ~/jdk24/open/src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Float16Consts.java 
26c26
< package jdk.internal.math;
---
> package jdk.incubator.vector;
28,30c28,30
< import static java.lang.Float16.MIN_EXPONENT;
< import static java.lang.Float16.PRECISION;
< import static java.lang.Float16.SIZE;
---
> import static jdk.incubator.vector.Float16.MIN_EXPONENT;
> import static jdk.incubator.vector.Float16.PRECISION;
> import static jdk.incubator.vector.Float16.SIZE;
37c37
< public class Float16Consts {
---
> class Float16Consts {

jddarcy commented:
$ diff test/jdk/java/lang/Float16/BasicFloat16ArithTests.java  ~/jdk24/test/jdk/jdk/incubator/vector/BasicFloat16ArithTests.java 
26c26,27
<  * @bug 8329817 8334432 8339076
---
>  * @bug 8329817 8334432 8339076 8341260
>  * @modules jdk.incubator.vector
30c31,32
< import static java.lang.Float16.*;
---
> import jdk.incubator.vector.Float16;
> import static jdk.incubator.vector.Float16.*;

$ diff test/jdk/java/math/BigDecimal/DoubleFloatValueTests.java  ~/jdk24/test/jdk/java/math/BigDecimal/DoubleFloatValueTests.java 
26c26
<  * @bug 8205592 8339252
---
>  * @bug 8205592 8339252 8341260
27a28
>  * @modules jdk.incubator.vector
37a39
> import jdk.incubator.vector.Float16;
110c112
<         Float16 res = bv.float16Value();
---
>         Float16 res =  Float16.valueOf(bv); // bv.float16Value();

jddarcy commented:
> ```
> $ diff test/jdk/java/lang/Float16/BasicFloat16ArithTests.java  ~/jdk24/test/jdk/jdk/incubator/vector/BasicFloat16ArithTests.java 
> 26c26,27
> <  * @bug 8329817 8334432 8339076
> ---
> >  * @bug 8329817 8334432 8339076 8341260
> >  * @modules jdk.incubator.vector
> 30c31,32
> < import static java.lang.Float16.*;
> ---
> > import jdk.incubator.vector.Float16;
> > import static jdk.incubator.vector.Float16.*;
> 
> $ diff test/jdk/java/math/BigDecimal/DoubleFloatValueTests.java  ~/jdk24/test/jdk/java/math/BigDecimal/DoubleFloatValueTests.java 
> 26c26
> <  * @bug 8205592 8339252
> ---
> >  * @bug 8205592 8339252 8341260
> 27a28
> >  * @modules jdk.incubator.vector
> 37a39
> > import jdk.incubator.vector.Float16;
> 110c112
> <         Float16 res = bv.float16Value();
> ---
> >         Float16 res =  Float16.valueOf(bv); // bv.float16Value();
> ```

Initially, I left the tests for the BigDecimal -> Float16 conversion in the tests for the base module to ease review compared to their location in the Valhalla branch. The tests can be extracted and moved to a jdk.incubator.vector area subsequently.

jddarcy commented:
> To ease review, diffs of corresponding files from the current, at time of writing, state of files in lworld+fp16 vs a port to jdk.incubator.vector, starting with Float16:
> 
> ```
> $ diff src/java.base/share/classes/java/lang/Float16.java  ~/jdk24/open/src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Float16.java 
> 26c26
> < package java.lang;
> ---
> > package jdk.incubator.vector;
> 28a29
> > import java.math.BigInteger;
> 30,31c31,32
> < import jdk.internal.math.*;
> < import jdk.internal.vm.annotation.IntrinsicCandidate;
> ---
> > // import jdk.internal.math.*;
> > // import jdk.internal.vm.annotation.IntrinsicCandidate;
> 37c38
> <  * The {@code Float16} is a primitive value class holding 16-bit data
> ---
> >  * The {@code Float16} is a class holding 16-bit data
> 46,48d46
> <  * <p>This is a <a href="https://openjdk.org/jeps/401">primitive value class</a> and its objects are
> <  * identity-less non-nullable value objects.
> <  *
> 52a51,56
> >  * <p>This is a <a href="{@docRoot}/java.base/java/lang/doc-files/ValueBased.html">value-based</a>
> >  * class; programmers should treat instances that are
> >  * {@linkplain #equals(Object) equal} as interchangeable and should not
> >  * use instances for synchronization, or unpredictable behavior may
> >  * occur. For example, in a future release, synchronization may fail.
> >  *
> 62,64d65
> <  *
> <  * @author Jatin Bhateja
> <  * @since 20.00
> 69,70c70,71
> < @jdk.internal.MigratedValueClass
> < @jdk.internal.ValueBased
> ---
> > // @jdk.internal.MigratedValueClass
> > //@jdk.internal.ValueBased
> 213c214,215
> <         return FloatToDecimal.toString(f16.floatValue());
> ---
> >         return Float.toString(f16.floatValue());
> >         // return FloatToDecimal.toString(f16.floatValue());
> 420d421
> <      * @see BigDecimal#float16Value()
> 423c424,539
> <         return v.float16Value();
> ---
> >         return BigDecimalConversion.float16Value(v);
> >     }
> > 
> >     private class BigDecimalConversion {
> >         /*
> >          * Let l = log_2(10).
> >          * Then, L < l < L + ulp(L) / 2, that is, L = roundTiesToEven(l).
> >          */
> >         private static final double L = 3.321928094887362;
> > 
> >         private static final int P_F16 = PRECISION;  // 11
> >         private static final int Q_MIN_F16 = MIN_EXPONENT - (P_F16 - 1);  // -24
> >         private static final int Q_MAX_F16 = MAX_EXPONENT - (P_F16 - 1);  // 5
> > 
> >         /**
> >          * Powers of 10 which can be represented exactly in {@code
> >          * Float16}.
> >          */
> >         private static final Float16[] FLOAT16_10_POW = {
> >             Float16.valueOf(1), Float16.valueOf(10), Float16.valueOf(100),
> >             Float16.valueOf(1_000), Float16.valueOf(10_000)
> >         };
> > 
> >         public static Float16 float16Value(BigDecimal bd) {
> > //             int scale = bd.scale();
> > //             BigInteger unscaledValue = bd.unscaledValue();
> > 
> > //              if (unscaledValue.abs().compareTo(BigInteger.valueOf(Long.MAX_VALUE)) <= 0) {
> > //                 long intCompact = bd.longValue();
> > //                 Float16 v = Float16.valueOf(intCompact);
> > //                 if (scale == 0) {
> > //                     return v;
> > //                 }
> > //                 /*
> > //                  * The discussion for the double case also applies here. That is,
> > //                  * the following test is precise for all long values, but here
> > //                  * Long.MAX_VALUE is not an issue.
> > //                  */
> > //                 if (v.longValue() == intCompact) {
> > //                     if (0 < scale && scale < FLOAT16_10_POW.length) {
> > //                         return Float16.divide(v, FLOAT16_10_POW[scale]);
> > //                     }
> > //                     if (0 > scale && scale > -FLOAT16_10_POW.length) {
> > //                         return Float16.multiply(v, FLOAT16_10_POW[-scale]);
> > //                     }
> > //                 }
> > //             }
> >             return fullFloat16Value(bd);
> >         }
> > 
> >         private static BigInteger bigTenToThe(int scale) {
> >             return BigInteger.TEN.pow(scale);
> >         }
> > 
> >         private static Float16 fullFloat16Value(BigDecimal bd) {
> >             if (BigDecimal.ZERO.compareTo(bd) == 0) {
> >                 return Float16.valueOf(0);
> >             }
> >             BigInteger w = bd.unscaledValue().abs();
> >             int scale = bd.scale();
> >             long qb = w.bitLength() - (long) Math.ceil(scale * L);
> >             Float16 signum = Float16.valueOf(bd.signum());
> >             if (qb < Q_MIN_F16 - 2) {  // qb < -26
> >                 return Float16.multiply(signum, Float16.valueOf(0));
> >             }
> >             if (qb > Q_MAX_F16 + P_F16 + 1) {  // qb > 17
> >                 return Float16.multiply(signum, Float16.POSITIVE_INFINITY);
> >             }
> >             if (scale < 0) {
> >                 return Float16.multiply(signum, valueOf(w.multiply(bigTenToThe(-scale))));
> >             }
> >             if (scale == 0) {
> >                 return Float16.multiply(signum, valueOf(w));
> >             }
> >             int ql = (int) qb - (P_F16 + 3);
> >             BigInteger pow10 =  bigTenToThe(scale);
> >             BigInteger m, n;
> >             if (ql <= 0) {
> >                 m = w.shiftLeft(-ql);
> >                 n = pow10;
> >             } else {
> >                 m = w;
> >                 n = pow10.shiftLeft(ql);
> >             }
> >             BigInteger[] qr = m.divideAndRemainder(n);
> >             /*
> >              * We have
> >              *      2^12 = 2^{P+1} <= i < 2^{P+5} = 2^16
> >              * Contrary to the double and float cases, where we use long and int, resp.,
> >              * here we cannot simply declare i as short, because P + 5 < Short.SIZE
> >              * fails to hold.
> >              * Using int is safe, though.
> >              *
> >              * Further, as Math.scalb(Float16) does not exists, we fall back to
> >              * Math.scalb(double).
> >              */
> >             int i = qr[0].intValue();
> >             int sb = qr[1].signum();
> >             int dq = (Integer.SIZE - (P_F16 + 2)) - Integer.numberOfLeadingZeros(i);
> >             int eq = (Q_MIN_F16 - 2) - ql;
> >             if (dq >= eq) {
> >                 return Float16.valueOf(bd.signum() * Math.scalb((double) (i | sb), ql));
> >             }
> >             int mask = (1 << eq) - 1;
> >             int j = i >> eq | (Integer.signum(i & mask)) | sb;
> >             return Float16.valueOf(bd.signum() * Math.scalb((double) j, Q_MIN_F16 - 2));
> >         }
> > 
> >         public static Float16 valueOf(BigInteger bi) {
> >             int signum = bi.signum();
> >             return (signum == 0 || bi.bitLength() <= 31)
> >                 ? Float16.valueOf(bi.longValue())  // might return infinities
> >                 : signum > 0
> >                 ? Float16.POSITIVE_INFINITY
> >                 : Float16.NEGATIVE_INFINITY;
> >         }
> 577,578c693
> <      * according to the IEEE 754 floating-point binary16 bit layout,
> <      * preserving Not-a-Number (NaN) values.
> ---
> >      * according to the IEEE 754 floating-point binary16 bit layout.
> 591,607d705
> <      * Returns a representation of the specified floating-point value
> <      * according to the IEEE 754 floating-point binary16 bit layout.
> <      *
> <      * @param   fp16   a {@code Float16} floating-point number.
> <      * @return the bits that represent the floating-point number.
> <      *
> <      * @see Float#floatToIntBits(float)
> <      * @see Double#doubleToLongBits(double)
> <      */
> <     public static short float16ToShortBits(Float16 fp16) {
> <         if (!isNaN(fp16)) {
> <             return float16ToRawShortBits(fp16);
> <         }
> <         return 0x7e00;
> <     }
> < 
> <     /**
> 694c792
> <     @IntrinsicCandidate
> ---
> >     // @IntrinsicCandidate
> 714c812
> <     @IntrinsicCandidate
> ---
> >     // @IntrinsicCandidate
> 783c881
> <     @IntrinsicCandidate
> ---
> >     // @IntrinsicCandidate
> 806c904
> <     @IntrinsicCandidate
> ---
> >     // @IntrinsicCandidate
> 829c927
> <     @IntrinsicCandidate
> ---
> >     // @IntrinsicCandidate
> 852c950
> <     @IntrinsicCandidate
> ---
> >     // @IntrinsicCandidate
> 873c971
> <     @IntrinsicCandidate
> ---
> >     // @IntrinsicCandidate
> 907c1005
> <     @IntrinsicCandidate
> ---
> >     // @IntrinsicCandidate
> 1109c1207
> <     @IntrinsicCandidate
> ---
> >     // @IntrinsicCandidate
> 1131c1229
> <     @IntrinsicCandidate
> ---
> >     // @IntrinsicCandidate
> 1319a1418
> >         int DoubleConsts_EXP_BIAS = 1023;
> 1328c1427
> <                 * Double.longBitsToDouble((long) (scaleFactor + DoubleConsts.EXP_BIAS) << Double.PRECISION - 1));
> ---
> >                 * Double.longBitsToDouble((long) (scaleFactor + DoubleConsts_EXP_BIAS) << Double.PRECISION - 1));
> 1330d1428
> < 
> 1374d1471
> < 
> ```

As initially done here, the port of Float16 omits any VM intrinsification; that would need to be done by subsequent work. Subsequent work may also adapt the Java implementations to be more amenable to being intrinsified.

rose00 commented:
Joe, our revised and now-current thinking about JIT support for the Float16 (both as a pre-Valhalla VBC and Valhalla value class) is that there should be zero need for any `@IntrinsicCandidate` markers in `Float16.java`.

The two existing intrinsics in `Float.java` should be fully sufficient to cue the JIT to introduce FP16 types and operations in its IR.  There is no need to help the JIT either recognize Float16 objects as special types, nor recognize the special interpretation of the `short` value field inside.  It is very routine in the Valhalla JIT to strip off the boxing (i.e., the VBC-like Float16 details) and just deal directly with the payload (i.e., the `short` value).  And it is also routine to change perspective on live values in the IR, depending on how they are used (i.e, via the two `Float` intrinsics), so that the IR ends up storing FP16 temps in whatever register class the CPU likes best.

So, unless there is some reason I am not aware of, there does not need to be any intrinsic definition in `Float16.java`, nor even a comment that gestures at the possible existence of an intrinsic.]

Note the economy of the resulting design:  Two intrinsics at the Java level provide the JIT with ample opportunity (if it so desires) to organize an IR with a full FP16 type and all its typed operations.  There is no need to surface such complexity in the Java source code.

rose00 commented:
Somebody might ask as a followup, "But what about calling sequences?  Without intrinsics, how does the JIT know to store Float16 values in the correct type of argument register?"   Presumably some platforms may wish to store arguments (and return values) in floating point registers, not in the integer registers used by a typical value class (such as Float16, containing a single `short` value).

That question might be motivated by the observation that, while FP16 stored in temps might well be stored in FPU registers, simply due to proximity to decode/encode intrinsics, changing the calling sequence requires that we give the Float16 class a little extra magic.  (More detail:  There are fundamentally 3 different places that FP16 values might need type-specific storage:  IR temps, argument/return registers, and heap variables.  The first can often be FPU regs.  The last can be untyped memory words and a `short` is just fine.  The middle are driven by the type of the value field, but can also take the field container Float16 into account.)

In any case, getting those details right does not require any `@IntrinsicCandidate` annotation, or any other source-level annotation.  It just requires the VM to know about Float16 and treat it a little bit specially.  Wholesale marking of methods is overkill.

rose00 commented:
Comparing with https://github.com/openjdk/jdk/pull/21490 we can see that there are more than minimum number of intrinsics I recommended above, but (crucially) the intrinsics are decoupled from the box type, and refactored into `Float16Math.java`.

https://github.com/openjdk/jdk/pull/21490/files#diff-105a2bf4929174c594b5bcc54f749e93d9fca1b5371ca301fff02badd0e8da5a

For example, see the max intrinsics line 53.  These intrinsics are better structured for the JIT, because there is no extraneous boxing to deal with.  (Boxes involve null checks and heap allocation and other details which are just extra overhead.)

Apart from intrinsic classification, a simpler thing to do that may help the box methods to fold up correctly would be a `@ForceInline` annotation.  But for the simple methods in this PR that is usually overkill.

jddarcy commented:
> Comparing with #21490 we can see that there are more than minimum number of intrinsics I recommended above, but (crucially) the intrinsics are decoupled from the box type, and refactored into `Float16Math.java`.
> 
> https://github.com/openjdk/jdk/pull/21490/files#diff-105a2bf4929174c594b5bcc54f749e93d9fca1b5371ca301fff02badd0e8da5a
> 
> For example, see the max intrinsics line 53. These intrinsics are better structured for the JIT, because there is no extraneous boxing to deal with. (Boxes involve null checks and heap allocation and other details which are just extra overhead.)
> 
> Apart from intrinsic classification, a simpler thing to do that may help the box methods to fold up correctly would be a `@ForceInline` annotation. But for the simple methods in this PR that is usually overkill.

I'm agnostic toward how the intrinsification of Float16 is implemented. I've removed the commented-out intrinsics annotations in a subsequent push. As noted, for review purposes I wanted to have a least one iteration of Float16 in the incubator that was as close as possible to the version of Float16 that has been in use in the lworld+fp16 branch.

jatin-bhateja commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Float16.java:
```diff
@@ -0,0 +1,1461 @@
+    public static Float16 sqrt(Float16 radicand) {
+        // Rounding path of sqrt(Float16 -> double) -> Float16 is fine
+        // for preserving the correct final value. The conversion
+        // Float16 -> double preserves the exact numerical value. The
+        // of the double -> Float16 conversion also benefits from the
```
```suggestion
        // square root of the double -> Float16 conversion also benefits from the
```

jatin-bhateja replied:
```suggestion
        // Float16 -> double preserves the exact numerical value. The
        // conversion of the double -> Float16 also benefits from the
```

jatin-bhateja commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Float16.java:
```diff
@@ -0,0 +1,1461 @@
+        // Float16 -> double preserves the exact numerical value. The
+        // of the double -> Float16 conversion also benefits from the
+        // 2p+2 property of IEEE 754 arithmetic.
+        return valueOf(Math.sqrt(radicand.doubleValue()));
```
Kindly also add a reference to link you shared on lworld+fp16 branch.
https://dl.acm.org/doi/pdf/10.1145/221332.221334

jddarcy replied:
That reference is already cited around line 850.

turbanoff commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Float16.java:
```diff
@@ -0,0 +1,1461 @@
+                return Float16.multiply(signum, valueOf(w));
+            }
+            int ql = (int) qb - (P_F16 + 3);
+            BigInteger pow10 =  bigTenToThe(scale);
```
```suggestion
            BigInteger pow10 = bigTenToThe(scale);
```

turbanoff commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Float16.java:
```diff
@@ -0,0 +1,1461 @@
+     */
+    /*package*/ static int getExponent0(short bits) {
+        // package private to be usable in java.lang.Float.
+        int bin16ExpBits     = 0x0000_7c00 & bits;     // Five exponent bits.
```
```suggestion
        int bin16ExpBits = 0x0000_7c00 & bits;     // Five exponent bits.
```

turbanoff commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Float16.java:
```diff
@@ -0,0 +1,1461 @@
+    public static Float16 ulp(Float16 f16) {
+        int exp = getExponent(f16);
+
+        return switch(exp) {
```
```suggestion
        return switch (exp) {
```

turbanoff commented on test/jdk/java/math/BigDecimal/DoubleFloatValueTests.java:
```diff
@@ -90,6 +108,15 @@ private static void checkFloat(BigDecimal bv, float exp) {
     }
 
+    private static void checkFloat16(BigDecimal bv, Float16 exp) {
+        Float16 res =  Float16.valueOf(bv); // bv.float16Value();
```
```suggestion
        Float16 res = Float16.valueOf(bv); // bv.float16Value();
```

rgiulietti commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Float16.java:
```diff
@@ -0,0 +1,1463 @@
+        };
+
+        public static Float16 float16Value(BigDecimal bd) {
+//             int scale = bd.scale();
+//             BigInteger unscaledValue = bd.unscaledValue();
+
+//              if (unscaledValue.abs().compareTo(BigInteger.valueOf(Long.MAX_VALUE)) <= 0) {
+//                 long intCompact = bd.longValue();
+//                 Float16 v = Float16.valueOf(intCompact);
+//                 if (scale == 0) {
+//                     return v;
+//                 }
+//                 /*
+//                  * The discussion for the double case also applies here. That is,
+//                  * the following test is precise for all long values, but here
+//                  * Long.MAX_VALUE is not an issue.
+//                  */
+//                 if (v.longValue() == intCompact) {
+//                     if (0 < scale && scale < FLOAT16_10_POW.length) {
+//                         return Float16.divide(v, FLOAT16_10_POW[scale]);
+//                     }
+//                     if (0 > scale && scale > -FLOAT16_10_POW.length) {
+//                         return Float16.multiply(v, FLOAT16_10_POW[-scale]);
+//                     }
+//                 }
+//             }
```
```suggestion
             int scale = bd.scale();
             BigInteger unscaledValue = bd.unscaledValue();

             if (unscaledValue.abs().compareTo(BigInteger.valueOf(Long.MAX_VALUE)) <= 0) {
                 long intCompact = unscaledValue.longValue();
                 Float16 v = Float16.valueOf(intCompact);
                 if (scale == 0) {
                     return v;
                 }
                 /*
                  * The discussion for the double case also applies here. That is,
                  * the following test is precise for all long values, but here
                  * Long.MAX_VALUE is not an issue.
                  */
                 if (v.longValue() == intCompact) {
                     if (0 < scale && scale < FLOAT16_10_POW.length) {
                         return Float16.divide(v, FLOAT16_10_POW[scale]);
                     }
                     if (0 > scale && scale > -FLOAT16_10_POW.length) {
                         return Float16.multiply(v, FLOAT16_10_POW[-scale]);
                     }
                 }
             }
```

jddarcy replied:
That does the trick; thanks for tracking down the mistake.

PaulSandoz commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Float16.java:
```diff
@@ -0,0 +1,1463 @@
+    }
+
+    // Skipping for now:
+    // public int hashCode()
+    // public static int hashCode(Float16 value)
+    // public boolean equals(Object obj)
```
I think we should add these methods in a subsequent PR. e.g., (although i may be ignorant of some details)
```
    @Override
    public int hashCode() {
        return Float16.hashCode(value);
    }

    public static int hashCode(float value) {
        return float16ToShortBits(value); 
    }

    public static int float16ToShortBits(float value) {
        if (!isNaN(value)) {
            return float16ToRawShortBits(value);
        }
        return Float16.NaN;
    }

    public boolean equals(Object obj) {
        return (obj instanceof Float16)
               && (float16ToShortBits(((Float16)obj).value) == float16ToShortBits(value));
    }
```

jddarcy replied:
Hmm. Yes, adding at least explicit hashCode/equals methods from Object is reasonable. Back when the type was in the Valhalla repo, I wasn't sure what the recommended practice regarding Object methods vis a vis JEP 401/402.

jddarcy replied:
Pushed equals/hashCode implementations. One slightly subtle point: the bit pattern for the "canonical" NaN is not defined. At runtime, the bit pattern of Float16.NaN is used for that purpose. Also, the exact bits returned for hashing is not defined, although I think it would be reasonable to add a requirement that "all distinct Float16 values have distinct hash codes."

PaulSandoz reviewed (commented):
Looking good, although i am not expert on the floating point aspects. Some of the more substantive comments could be addressed if chosen to in subsequent PRs.

jddarcy commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Float16.java:
```diff
@@ -0,0 +1,1463 @@
+     * 754.
+     *
+     * @param f16 the value to be negated
+     * @jls 15.15.4 Unary Minus Operator {@code -}
```
I'll add more precise tags as well.

jddarcy commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Float16.java:
```diff
@@ -0,0 +1,1463 @@
+// Currently Float16 is a value based class but in future will be aligned with
+// Enhanced Primitive Boxes described by JEP-402 (https://openjdk.org/jeps/402)
+// @jdk.internal.MigratedValueClass
+//@jdk.internal.ValueBased
```
Will update to refer to JEP 401.

jddarcy commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Float16.java:
```diff
@@ -0,0 +1,1463 @@
+    extends Number
+    implements Comparable<Float16> {
+    private final short value;
+    private static final long serialVersionUID = 16; // Not needed for a value class?
```
For now, I think the svuid is fine; we can update to whatever the right answer will be under value classes at that time. Likewise, leaving the Number superclass is fine for now, but may change later.

liach commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Float16.java:
```diff
@@ -0,0 +1,1518 @@
+     * @param  s the string to be parsed.
+     * @return the {@code Float16} value represented by the string
+     *         argument.
+     * @throws NullPointerException  if the string is null
```
Should we move this NPE clause to the class specification, i.e. `Unless otherwise specified, all methods in this class throw {@code NullPointerException} when passed a {@code null}`? I think this applies to the APIs taking `BigDecimal` and static methods taking `Float16`. (Note we need to explicitly exclude `equals`)

jddarcy replied:
Hmm. I'm trying to keep parity between Float16 and java.lang.{Float, Double}. Float/Double don't include the class-level statement so I wouldn't include on in Float16 until they do. Writing Float16 has prompted several other improvements to Float/Double, so a null-handling statement is something worth considering.

liach commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Float16.java:
```diff
@@ -0,0 +1,1518 @@
+     */
+    public static Float16 min(Float16 a, Float16 b) {
+        return shortBitsToFloat16(floatToFloat16(Math.min(a.floatValue(),
+                                                          b.floatValue()) ));
```
I assume we will optimize these min/max implementations in the future. Otherwise, the extra space should be removed.

jddarcy replied:
The general intention is that the back-end intrinsification efforts that use native hardware instructions, where present, will provide the main speed boost. Therefore, the preference has been to keep the Java sources in Float16 as straightforward and "obviously correct" as possible.

rgiulietti commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Float16.java:
```diff
@@ -0,0 +1,1518 @@
+                         return Float16.multiply(v, FLOAT16_10_POW[-scale]);
+                     }
+                 }
+             }
```
(nit)
Then indents are not "multiples of 4" in the lines of this method's body.

rgiulietti commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Float16.java:
```diff
@@ -0,0 +1,1518 @@
+     */
+    public static int hashCode(Float16 value) {
+        Float16 f16 = isNaN(value) ? Float16.NaN : value;
+        return (int)float16ToRawShortBits(f16);
```
What about randomizing the returned value? IIUC, JEP 401 does not speak against.
The purpose would not be to conceal anything, but to ensure that clients cannot rely on specific values.

```
    private static final int HASH_MASK = RandomGenerator.getDefault().nextInt();
    
    /**
     * Returns a hash code for a {@code Float16} value; compatible with
     * {@code Float16.hashCode()}.
     *
     * @param value the value to hash
     * @return a hash code value for a {@code Float16} value.
     */
    public static int hashCode(Float16 value) {
        Float16 f16 = isNaN(value) ? Float16.NaN : value;
        int bits = float16ToRawShortBits(f16);
        return (bits << 16 | bits & 0xFFFF) ^ HASH_MASK;
    }
```

jddarcy replied:
Hmm. Certainly at this point, I don't think hash randomization is necessary; although it might be helpful at a later point. Set.of() uses randomized iteration order as a precedent in that direction.

jddarcy commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Float16.java:
```diff
@@ -0,0 +1,1518 @@
+ * The methods in this class generally have analogous methods in
+ * either {@link Float}/{@link Double} or {@link Math}/{@link
+ * StrictMath}. Unless otherwise specified, the handling of special
+ * floating-point values such as {@linkplain #isNaN(Float16) NaN}
```
Reasonable suggestion; I'll put the same kind of pointers into java.lang.Double sections as java.lang.Float has.

jddarcy commented:
Updated the PR with new code and tests for properly rounded String -> Float16 conversion only using public JDK/Java SE APIs. Whenever Float16 support is moved to the base module, this implementation can be abandoned in favor of augmenting the existing support for String -> float and String -> double conversion.

jddarcy commented:
I believe the only remaining functionality to add is porting over the minimal-length Float16 to String conversion from JDK-8341245 in the Valhalla repo.

rgiulietti commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Float16.java:
```diff
@@ -0,0 +1,1667 @@
+        // characters rather than codepoints.
+
+        if (trialResult == 0.0 // handles signed zeros
+            || Math.abs(trialResult) > (65504.0 + 32.0) || // Float.MAX_VALUE + ulp(MAX_VALUE)
```
```suggestion
            || Math.abs(trialResult) > (65504.0 + 16.0) || // Float.MAX_VALUE + ulp(MAX_VALUE) / 2
```

jddarcy replied:
When writing the code, I didn't think through all the possible boundary conditions near Float16.MAX_VALUE + 0.5*ulp(Float16.MAX_VALUE) so I bumped out the overflow limit.

rgiulietti commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Float16.java:
```diff
@@ -0,0 +1,1667 @@
+                if (periodIndex != -1) {
+                    // Reconstruct integer and fraction digit sequence
+                    // without the period.
+                    hexSignificand.append(s.substring(digitStart,      periodIndex));
+                    hexSignificand.append(s.substring(periodIndex + 1, pIndex));
```
```suggestion
                    hexSignificand.append(s, digitStart, periodIndex);
                    hexSignificand.append(s, periodIndex + 1, pIndex);
```

jddarcy replied:
Good refactoring; thanks for the suggestion.

jddarcy commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Float16.java:
```diff
@@ -0,0 +1,1667 @@
+        // be skipped since the double-rounding would be known not to
+        // have occurred.
+        long dAsLong = Double.doubleToRawLongBits(d);
+        long mask = 0x01FF_FFFF_FFFFL; // 41 low-order bits
+        long maskedValue = dAsLong & mask;
+        // Can't have all-zeros or all-ones in low-order bits
+        return maskedValue != 0L && maskedValue != mask;
```
Let's talk about the case analysis off-list.

rgiulietti commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Float16.java:
```diff
@@ -0,0 +1,1667 @@
+     *               parsable {@code Float16}.
+     * @see    java.lang.Float#valueOf(String)
+     */
+    public static Float16 valueOf(String s) throws NumberFormatException {
```
The current implementation throws when the input is `"NaN"`, with or without an optional sign.

jddarcy replied:
Thanks for catching that; fix and tests added in subsequent push.

rgiulietti replied:
Single digit inputs, e.g., `Float16.valueOf"2")` throw as well.

jddarcy replied:
Addressed in most recent push; thanks.

rgiulietti commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Float16.java:
```diff
@@ -0,0 +1,2345 @@
+     *
+     * The behavior of this class is analogous to {@link
+     * Float#toHexString(float)} except that an exponent value of
+     * {@code "p14"} is used for subnormal {@code Float16} values.
```
```suggestion
     * {@code "p-14"} is used for subnormal {@code Float16} values.
```

rgiulietti commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Float16.java:
```diff
@@ -0,0 +1,2345 @@
+     * Returns the larger of two {@code Float16} values.
+     *
+     * The handling of signed zeros, NaNs, infinities, and other
+     * special cases by this method is analagous to the handling of
```
```suggestion
     * special cases by this method is analogous to the handling of
```

rgiulietti commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Float16.java:
```diff
@@ -0,0 +1,2345 @@
+     * Returns the smaller of two {@code Float16} values.
+     *
+     * The handling of signed zeros, NaNs, infinities, and other
+     * special cases by this method is analagous to the handling of
```
```suggestion
     * special cases by this method is analogous to the handling of
```

rgiulietti commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Float16.java:
```diff
@@ -0,0 +1,2345 @@
+     * @see    java.lang.Float#valueOf(String)
+     */
+    public static Float16 valueOf(String s) throws NumberFormatException {
+        s = s.trim(); // Legacy behavior from analagous methods on
```
```suggestion
        s = s.trim(); // Legacy behavior from analogous methods on
```

rgiulietti reviewed (approved):
Looks good.
Approving the current status.
Are there plans for other small additions?

jddarcy commented:
> Looks good. Approving the current status. Are there plans for other small additions?

Subsequent to this comment, I did a light clean-up pass over the tests. I'll do a pass over the Float16 implementation too to make sure previous comments are all addressed.

jddarcy commented:
> > Looks good. Approving the current status. Are there plans for other small additions?
> 
> Subsequent to this comment, I did a light clean-up pass over the tests. I'll do a pass over the Float16 implementation too to make sure previous comments are all addressed.

PS Pushed another clean-up pass over the implementation and tests.

jddarcy commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Float16.java:
```diff
@@ -0,0 +1,1667 @@
+// expected to be aligned with Value Classes and Object as described in
+// JEP-401 (https://openjdk.org/jeps/401).
+// @jdk.internal.MigratedValueClass
+// @jdk.internal.ValueBased
```
The jdk.internal.ValueBased annotation is not exported for use outside of the java.base module.

PaulSandoz replied:
We can do this in a subsequent PR and export the package `jdk.internal` to module `jdk.incubating.vector` like we do for other `java.base` internal packages. I believe no other changes should be necessary, but we should double check.

Classes in the `jdk.incubating.vector` module are loaded in the boot class loader and are considered privileged, therefore the internal `@ValueBased` annotation will not be ignored by the runtime.

rgiulietti reviewed (approved):
LGTM

rgiulietti commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Float16.java:
```diff
@@ -0,0 +1,2353 @@
+    /**
+     * A constant holding the largest positive finite value of type
+     * {@code Float16},
+     * (2-2<sup>-10</sup>)&middot;2<sup>15</sup>, numerically equal to 65504.0.
```
Not sure if this is the reason, but in decimal notation `MIN_NORMAL` and `MIN_VALUE` are relatively unreadable: `0.00006103515625` and `5.9604644775390625E-8`, resp.

jddarcy replied:
Yes, that was the reason; the other constants written out aren't particularly informative to most users.
